core/src/main/java/com/radixdlt/consensus/liveness/MultiFactorPacemakerTimeoutCalculator.java

```diff
@@ -65,57 +65,61 @@
-  private final long baseTimeoutMilliseconds;
-  private final double rate;
-  private final int maxExponent;
-  private final long additionalRoundTimeIfProposalReceivedMs;
+public final class MultiFactorPacemakerTimeoutCalculator implements PacemakerTimeoutCalculator {
+  private final PacemakerTimeoutCalculatorConfig config;
 
   @Inject
-  public ExponentialPacemakerTimeoutCalculator(
-      @PacemakerBaseTimeoutMs long baseTimeoutMilliseconds,
-      @PacemakerBackoffRate double rate,
-      @PacemakerMaxExponent int maxExponent,
-      @AdditionalRoundTimeIfProposalReceivedMs long additionalRoundTimeIfProposalReceivedMs) {
-    if (baseTimeoutMilliseconds <= 0) {
-      throw new IllegalArgumentException(
-          "timeoutMilliseconds must be > 0 but was " + baseTimeoutMilliseconds);
-    }
-    if (rate <= 1.0) {
-      throw new IllegalArgumentException("rate must be > 1.0, but was " + rate);
-    }
-    if (maxExponent < 0) {
-      throw new IllegalArgumentException("maxExponent must be >= 0, but was " + maxExponent);
-    }
-    double maxTimeout = baseTimeoutMilliseconds * Math.pow(rate, maxExponent);
-    if (maxTimeout > Long.MAX_VALUE) {
-      throw new IllegalArgumentException(
-          "Maximum timeout value of " + maxTimeout + " is too large");
-    }
-
-    this.baseTimeoutMilliseconds = baseTimeoutMilliseconds;
-    this.rate = rate;
-    this.maxExponent = maxExponent;
-    this.additionalRoundTimeIfProposalReceivedMs = additionalRoundTimeIfProposalReceivedMs;
+  public MultiFactorPacemakerTimeoutCalculator(PacemakerTimeoutCalculatorConfig config) {
+    this.config = config;
   }
 
   @Override
-  public long calculateTimeoutMs(long timeoutOccurrences) {
-    double exponential = Math.pow(this.rate, Math.min(this.maxExponent, timeoutOccurrences));
-    return Math.round(this.baseTimeoutMilliseconds * exponential);
+  public long calculateTimeoutMs(long timeoutOccurrences, double vertexStoreUtilizationRatio) {
+    final var exponential =
+        Math.pow(config.rate(), Math.min(config.maxExponent(), timeoutOccurrences));
+
+    final var exponentialTimeout = config.baseTimeoutMs() * exponential;
+
+    // It should already be in the [0, 1] range, but we're nonetheless sanitizing the input
+    final var vertexStoreUtilizationRatioClamped =
+        Math.max(0, Math.min(1, vertexStoreUtilizationRatio));
+
+    // We're only applying the multiplier if vertexStoreUtilizationRatio is
+    // on or above vertexStoreMultiplierThreshold: we're translating from
+    // range [vertexStoreMultiplierThreshold, 1] to [1, maxVertexStoreMultiplier].
+    // The multiplier starts at 1 right at the threshold
+    // and linearly grows to reach maxVertexStoreMultiplier
+    // when vertexStoreUtilizationRatio = 1.
+    final var multiplier =
+        Math.max(
+            1, // Multiplier is 1 (i.e. no-op) if we're below the threshold
+            lerp(
```

Comment:
guess what... Guava has the lerp function, with a cool (if not too-fluent) style: `LinearTransformation.mapping(0.66, 1.0).and(1.0, 10.0).transform(ratio);`

Reply:
guava has it all :smile: